nutils/evaluable.py

```diff
@@ -757,15 +758,34 @@ def _assparse(self):
       return chunks
     return _assparse
 
-  class _ArrayMeta(type(Evaluable)):
+  class _ArrayMeta(_ArrayMeta):
     def __new__(mcls, name, bases, namespace):
       if '_assparse' in namespace:
         namespace['_assparse'] = _chunked_assparse_checker(namespace['_assparse'])
       return super().__new__(mcls, name, bases, namespace)
 
-else:
-
-  _ArrayMeta = type(Evaluable)
+if debug_flags.evalf:
+  class _evalf_checker:
+    def __init__(self, orig):
+      self.evalf_obj = getattr(orig, '__get__', lambda *args: orig)
+    def __get__(self, instance, owner):
+      evalf = self.evalf_obj(instance, owner)
+      @functools.wraps(evalf)
+      def evalf_with_check(*args, **kwargs):
+        res = evalf(*args, **kwargs)
+        if hasattr(instance, 'dtype'):
+          assert asdtype(res.dtype) == instance.dtype, ((instance.dtype, res.dtype), instance, res)
+          if hasattr(instance, 'ndim'):
+            assert res.ndim == instance.ndim
+          assert all(m == n for m, n in zip(res.shape, instance.shape) if isinstance(n, int)), 'shape mismatch'
```

Comment:
If there is no `ndim`, then there should not be a `shape` either. Is the `if hasattr(instance, 'ndim')` test necessary?

I think not having a `dtype` doesn't imply there isn't a shape. Maybe we should place the `ndim` and `shape` tests under a separate `if hasattr(instance, 'shape')`?

Reply:
What do you think of the following:

```python
assert not hasattr(instance, 'dtype') or asdtype(res.dtype) == instance.dtype, ((instance.dtype, res.dtype), instance, res)
assert not hasattr(instance, 'ndim') or res.ndim == instance.ndim
assert not hasattr(instance, 'shape') or all(m == n for m, n in zip(res.shape, instance.shape) if isinstance(n, int)), 'shape mismatch'
```

Reply:
Done. Thanks!